### ISSUE TITLE
WebSockets support

### ISSUE THREAD

**Issue by florimondmanca** [bocadilloproject/bocadillo#75]
**Is your feature request related to a problem? Please describe.**
WebSockets enable a whole set of real-time web applications, which look like great candidates for async web servers. It should be as easy to build a backend websocket server than it is to create a websocket client in the frontend world.

**Describe the solution you'd like**
Implement a friendly, async/await-based interface for building websocket views, which would allow to write something like:

```python
from bocadillo import API, WebSocket, websocket_view

api = API()

@api.websocket_route("/chat")
async def chat(ws: WebSocket):
    # The connection has been accepted.
    # v Shortcut for a `while True` loop that calls `receive_<data_type>()`
    async for message in ws.messages(data_type="text"):
        # Process message… here, just echo back to the client
        await ws.send({"type": "echo", "data": message})
    # Connection automatically closes when leaving this view,
    # or if an exception occurs.
```

For the implementation, this feature should mostly be a wrapper around [Starlette WebSockets](https://www.starlette.io/websockets/).

**Describe alternatives you've considered**
- We could require that the developer manages calls to `accept` and `close` themselves, but that would involve too much boilerplate.
- A WS equivalent of hooks is not compatible with the proposed API, because hooks will have to run *before* the connection is accepted and *after* it is closed (and how about after_connect and before_close hooks?). That seems to add too much complexity. A solution could be a `manual: bool` flag defaulting to `False` which, if `True`, does not automatically accept/close the connection. Usage will have to be documented.

**Additional context**
Came to this conclusion while thinking about example projects for a tutorial. :)

**Comment by florimondmanca** [comment]
@cs01 Thanks for the heads up! I was interested in that thumbsup you left on the issue, my questions are now answered. ;-) Looking forward to seeing Bocadillo + WebSockets in action in pyxterm.js once we roll this feature out. :+1:

**Comment by cs01** [comment]
In the back of my mind I was thinking of using it on https://github.com/cs01/gdbgui, but that would be a much bigger undertaking. I realized today that a good first step would be pyxterm.js.

**Comment by florimondmanca** [comment]
Yep, that's a smart move. Better to start with a proof of concept than putting the stability of a 6k+ star project at risk.

**Comment by cs01** [comment]
Great work, can’t wait to try it out.

**Comment by florimondmanca** [comment]
Update: v0.9 has just been released :wink: https://pypi.org/project/bocadillo/0.9.0
